docs/design-notes/qualifying-replicas-for-reads.md

```diff
@@ -0,0 +1,90 @@
+# Qualifying replicas for read operations
+
+## Failure detection based on last seen information
+
+### Intro
+
+When a read request arrives at the coordinator node, it is later
+forwarded to one or more replicas in order to read the actual data
+and return it back to the client. Each request has an associated
+timeout, which indicates how long the client is willing to wait
+for the query results.
+This timeout information can be leveraged to lift some pressure off
+the coordinator if some of the target replicas are in fact dead,
+having network problems, etc.
+
+### Gathering last seen information
+
+The coordinator should be able to predict whether a replica is likely
+to respond to a request in time or not. This prediction should be as
+precise as possible, but at the same time have minimal overhead, because
+performing read requests is on the hot path. A simple failure detection
+mechanism with high granularity is implemented as follows:
+ * each time a request is sent to a replica, this attempt's timestamp
+   is recorded
+ * first timestamp of a streak of unresponded requests is recorded separately
+   in order to see how long ago the coordinator started to try to communicate
+   with a replica; example:
+    1. sent request A
+    2. received a response for A
+    3. sent request B <-- this is the interesting timestamp, since the node
+         hasn't responded at all after this request was sent
+    4. sent request C
+    5. sent request D
+ * each time a response is received from a replica, the node is marked as responsive
+
+The "last time a request was sent without response" metrics is a cheap, rough
+implementation of a failure detector - if the node hasn't responded for more than
+X milliseconds since the previous request and the current request is going
+to time out in X milliseconds, the target node gets more and more likely to be
+qualified as dead.
+
+Finally, when a replica was not contacted in a long while (e.g. the full timeout period),
+a communication attempt will be performed anyway, just in case the replica came back
+to full responsiveness.
+
+### Omitting replicas not likely to respond in time
+
+Based on the last seen information, the coordinator is now allowed
+to make an educated guess whether a replica is alive and ready to respond.
+A simple, probabilistic approach is implemented as follows.
+At first, we compute how much time left (`TL`) the request still has until
+it reaches its deadline and times out. From the last seen information
+we fetch the last time without any response: `TWR`.
+If `TL` is larger than `TWR`, then the node is likely alive, so the request
```

Comment:
I'm not entirely convinced about this idea either - I just liked the fact that **if we assume** that interactive workloads use small timeout values and batch workloads use large timeout values, we'll get more aggressive failure detection for interactive workloads. That's because I treat a small timeout as an SLA expectation - "I need this query to finish in 200ms or I don't care about the result anymore". On the other hand, long timeout values suggest that the request can take its time to finish, and it's more important to finish it eventually that to kill it and have better latency for other similar requests.

On the other hand, we already have another mechanism for specifying the workload type - users can set it up via service levels. So perhaps the more aggressive failure detection can only be activated for interactive workloads.

Finally, if we don't use `TL`, do you have another suggestion what to base the computations on? We still need a value to decide what's the sensitivity of this failure detector. Some alternatives are
 - a hardcoded value that we deem reasonable, e.g. "5 seconds"
 - a configuration variable (perhaps per-service-level)

Suggestions/ideas?

Reply:
*discussion ping*